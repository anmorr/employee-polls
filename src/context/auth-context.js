
import React, { useState } from 'react';

import {
	CognitoUserPool
} from 'amazon-cognito-identity-js';

const authData = {
    userPoolId: 'us-east-1_EhgVSlUt4', // e.g. us-east-2_uXboG5pAb
    userPoolClientId: '1agnserutlvfh3811qresi5ivc', // e.g. 25ddkmj4v6hfsfvruhpfi7n4hv
    region: 'us-east-1' // e.g. us-east-2
  }

var poolData = {
    UserPoolId: authData.userPoolId,
    ClientId: authData.userPoolClientId
}

var userPool = new CognitoUserPool(poolData);

const AuthContext = React.createContext({
    // Define the shape of our context
    // which also allows better autocompletion
    token: '',
    isLoggedIn: false,
    login: (token, expirationTime) => {},
    logout: () => {},
    authData: {},
    userPool: '',
    congitoUser: ''
});


export const AuthContextProvider = (props) => {

    
    let initialToken = '';
    let cognitoUser = userPool.getCurrentUser();

    if (cognitoUser != null) {
        cognitoUser.getSession(function(err, session) {
            if (err) {
                alert(err.message || JSON.stringify(err));
                return;
            }
            // console.log('session validity: ' + session.isValid());
            initialToken = session.getIdToken().getJwtToken()

            // NOTE: getSession must be called to authenticate user before calling getUserAttributes
            cognitoUser.getUserAttributes(function(err, attributes) {
                if (err) {
                    // Handle error
                } else {
                    // Do something with attributes
                    // console.log(attributes)
                }
            });

            // AWS.config.credentials = new AWS.CognitoIdentityCredentials({
            //     IdentityPoolId: '...', // your identity pool id here
            //     Logins: {
            //         // Change the key below according to the specific region your user pool is in.
            //         'cognito-idp.<region>.amazonaws.com/<YOUR_USER_POOL_ID>': session
            //             .getIdToken()
            //             .getJwtToken(),
            //     },
            // });

            // Instantiate aws sdk service objects now that the credentials have been updated.
            // example: var s3 = new AWS.S3();
        });
    }

    // console.log('initialToken: ', initialToken)
    const [token, setToken] = useState(initialToken);

    const userIsLoggedIn = !!token; 

    const logoutHandler = () => {
        setToken(null);
        userPool.getCurrentUser().signOut();
    }

    const loginHandler = (token) => {
        setToken(token);
    };


    const contextValue = {
        token: token,
        isLoggedIn: userIsLoggedIn,
        login: loginHandler,
        logout: logoutHandler,
        authData: authData,
        userPool: userPool,
        cognitoUser: cognitoUser,
    }

    return <AuthContext.Provider value={contextValue}>
                {props.children}
            </AuthContext.Provider>
}

export default AuthContext;