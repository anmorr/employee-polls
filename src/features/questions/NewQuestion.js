import { useHistory } from "react-router-dom";
import { useTextField } from "../../shared/hooks/hooks";
import { createTheme, ThemeProvider } from '@mui/material/styles';
import { addQuestion } from "./questionsSlice";
import { useDispatch } from "react-redux";
import { useState } from "react";
import { useContext } from "react";
import AuthContext from "../../context/auth-context";
import { addQuestionToUser } from "../users/usersSlice";
import QuestionCard from "./QuestionCard";
import { Alert } from "@mui/material";
import { Box } from "@mui/material";


export default function NewQuestion() {

    const history = useHistory();
    const dispatch = useDispatch();

    const authCtx = useContext(AuthContext);

    const currentUser = (
        authCtx.userPool.getCurrentUser() ?
            authCtx.userPool.getCurrentUser().username :
            'sarahedo'
    )
    
    const [optionOneProps, setOptionOne] = useTextField('');
    const [optionTwoProps, setOptionTwo] = useTextField('');
    const [addRequestStatus, setAddRequestStatus] = useState('idle')

    const [questionAdded, setQuestionAdded] = useState(false)


    const handleSubmit = async (event) => {
        event.preventDefault();
        if (addRequestStatus === 'idle') {
            try {
                setAddRequestStatus('pending')
                let optionOne = optionOneProps.value
                let optionTwo = optionTwoProps.value
                const newQuestion = {
                    optionOneText: optionOne,
                    optionTwoText: optionTwo,
                    author: currentUser,
                }
                const response = await dispatch(addQuestion(newQuestion)).unwrap()
                // console.log(response)
                dispatch(addQuestionToUser({
                    id: response.id,
                    author: response.author,
                }))
                setOptionOne('')
                setOptionTwo('')
                setQuestionAdded(true)
                // console.log('location: ', history.location)
                // history.push('/')
                // console.log('location: ', history.location)
                showSuccessTimeout(3000)
            } catch (err) {
                console.error('Failed to save the question: ', err)
            } finally {
                setAddRequestStatus('idle')
            }
        }
    }

    const showSuccessTimeout = (time) => {
        setTimeout(() => {
            setQuestionAdded(false)
            history.push('/')
        }, time)
    }
    
    const questionAddedSuccessAlert = (
        <Alert
            severity="success"
            inputProps={{
                "data-testid": "success-alert"
            }}
        >
            {/* <AlertTitle>Success</AlertTitle> */}
            New question has been added!
            You will now be redirected to the questions page...
        </Alert>
    )

    return (
        <Box>
            {questionAdded && questionAddedSuccessAlert}   
           
            {!questionAdded && <QuestionCard
                optionOneProps={optionOneProps}
                optionTwoProps={optionTwoProps}
                submitHandler={handleSubmit} />}
        </Box>
    )
}