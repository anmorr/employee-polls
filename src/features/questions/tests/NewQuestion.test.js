import {
    fireEvent,
    render, screen
} from '@testing-library/react';

import { AuthContextProvider } from '../../../context/auth-context';
import { store } from '../../../app/store';
import { Provider } from 'react-redux';
import NewQuestion from '../NewQuestion';


describe('New Question Page', () => {

    it('matches the snapshot', () => {
        var component = render(
            <Provider store={store}>
                    <AuthContextProvider>
                        <NewQuestion />
                    </AuthContextProvider>
            </Provider>
        );
        expect(component).toMatchSnapshot();
    });

    it('will display a success message if the question is added.', () => {

        var component = render(
                <Provider store={store}>
                    <AuthContextProvider>
                        <NewQuestion />
                    </AuthContextProvider>
            </Provider>);
        
        var optionOneInput = component.getByTestId('option-one')
        fireEvent.change(optionOneInput, { target: {value: 'Go to LA'}})
        var optionTwoInput = component.getByTestId('option-two')
        fireEvent.change(optionTwoInput, { target: {value: 'Go to MIA'}})
        var submitButton = component.getByTestId('submit-button')
        fireEvent.click(submitButton)

        setTimeout(function() {
            //your code to be executed after 1 second
            expect(component.getByTestId('success-alert')).toBeInTheDocument();
          }, 1500);
    })
})