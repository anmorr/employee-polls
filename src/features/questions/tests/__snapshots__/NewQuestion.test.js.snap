// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`New Question Page matches the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiBox-root css-0"
      >
        <main
          class="MuiContainer-root MuiContainer-maxWidthXs css-nzinn5-MuiContainer-root"
        >
          <div
            class="MuiBox-root css-bur7ie"
          >
            <h1>
              Would You Rather
            </h1>
            <h5
              class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
            >
              Create Your Own Poll
            </h5>
            <form
              class="MuiBox-root css-164r41r"
              novalidate=""
            >
              <h5
                class="MuiTypography-root MuiTypography-h5 css-ph2xsj-MuiTypography-root"
              >
                First Option
              </h5>
              <div
                class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
              >
                <label
                  class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="false"
                  for="optionOne"
                  id="optionOne-label"
                >
                  Option One
                  <span
                    aria-hidden="true"
                    class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                  >
                     
                    *
                  </span>
                </label>
                <div
                  class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline css-8ewcdo-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <textarea
                    aria-invalid="false"
                    class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                    data-testid="option-one"
                    id="optionOne"
                    name="optionOne"
                    required=""
                    rows="4"
                    style="height: 0px; overflow: hidden;"
                  />
                  <textarea
                    aria-hidden="true"
                    class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                    readonly=""
                    style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding: 0px; width: 100%;"
                    tabindex="-1"
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-1ftyaf0"
                    >
                      <span>
                        Option One
                         
                        *
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
              <h5
                class="MuiTypography-root MuiTypography-h5 css-ph2xsj-MuiTypography-root"
              >
                Second Option
              </h5>
              <div
                class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
              >
                <label
                  class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="false"
                  for="optionTwo"
                  id="optionTwo-label"
                >
                  Option Two
                  <span
                    aria-hidden="true"
                    class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                  >
                     
                    *
                  </span>
                </label>
                <div
                  class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline css-8ewcdo-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <textarea
                    aria-invalid="false"
                    class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                    data-testid="option-two"
                    id="optionTwo"
                    name="optionTwo"
                    required=""
                    rows="4"
                    style="height: 0px; overflow: hidden;"
                  />
                  <textarea
                    aria-hidden="true"
                    class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                    readonly=""
                    style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding: 0px; width: 100%;"
                    tabindex="-1"
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-1ftyaf0"
                    >
                      <span>
                        Option Two
                         
                        *
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
              <button
                class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButtonBase-root Mui-disabled  css-1vhaqj4-MuiButtonBase-root-MuiButton-root"
                data-testid="submit-button"
                disabled=""
                tabindex="-1"
                type="submit"
              >
                Submit
              </button>
            </form>
          </div>
          <p
            class="MuiTypography-root MuiTypography-body2 MuiTypography-alignCenter css-ijul54-MuiTypography-root"
          >
            Copyright © 
            <a
              class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1cghntr-MuiTypography-root-MuiLink-root"
              href="#"
            >
              Employee Polls
            </a>
             
            2022
            .
          </p>
        </main>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiBox-root css-0"
    >
      <main
        class="MuiContainer-root MuiContainer-maxWidthXs css-nzinn5-MuiContainer-root"
      >
        <div
          class="MuiBox-root css-bur7ie"
        >
          <h1>
            Would You Rather
          </h1>
          <h5
            class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
          >
            Create Your Own Poll
          </h5>
          <form
            class="MuiBox-root css-164r41r"
            novalidate=""
          >
            <h5
              class="MuiTypography-root MuiTypography-h5 css-ph2xsj-MuiTypography-root"
            >
              First Option
            </h5>
            <div
              class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for="optionOne"
                id="optionOne-label"
              >
                Option One
                <span
                  aria-hidden="true"
                  class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                >
                   
                  *
                </span>
              </label>
              <div
                class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline css-8ewcdo-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <textarea
                  aria-invalid="false"
                  class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                  data-testid="option-one"
                  id="optionOne"
                  name="optionOne"
                  required=""
                  rows="4"
                  style="height: 0px; overflow: hidden;"
                />
                <textarea
                  aria-hidden="true"
                  class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                  readonly=""
                  style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding: 0px; width: 100%;"
                  tabindex="-1"
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-1ftyaf0"
                  >
                    <span>
                      Option One
                       
                      *
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <h5
              class="MuiTypography-root MuiTypography-h5 css-ph2xsj-MuiTypography-root"
            >
              Second Option
            </h5>
            <div
              class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for="optionTwo"
                id="optionTwo-label"
              >
                Option Two
                <span
                  aria-hidden="true"
                  class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                >
                   
                  *
                </span>
              </label>
              <div
                class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline css-8ewcdo-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <textarea
                  aria-invalid="false"
                  class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                  data-testid="option-two"
                  id="optionTwo"
                  name="optionTwo"
                  required=""
                  rows="4"
                  style="height: 0px; overflow: hidden;"
                />
                <textarea
                  aria-hidden="true"
                  class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                  readonly=""
                  style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding: 0px; width: 100%;"
                  tabindex="-1"
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-1ftyaf0"
                  >
                    <span>
                      Option Two
                       
                      *
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <button
              class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButtonBase-root Mui-disabled  css-1vhaqj4-MuiButtonBase-root-MuiButton-root"
              data-testid="submit-button"
              disabled=""
              tabindex="-1"
              type="submit"
            >
              Submit
            </button>
          </form>
        </div>
        <p
          class="MuiTypography-root MuiTypography-body2 MuiTypography-alignCenter css-ijul54-MuiTypography-root"
        >
          Copyright © 
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1cghntr-MuiTypography-root-MuiLink-root"
            href="#"
          >
            Employee Polls
          </a>
           
          2022
          .
        </p>
      </main>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
